### PR TITLE
bugfix: refresh token when within refresh period or later

### PR DESCRIPTION
**Pull request by norrs**
Pulled in net.morimekta.utils.testing for easier testing for advancing a (fake) clock.
Review commit by commit.

### CODE REVIEW

File: java/src/test/java/com/wgtwo/auth/ClientCredentialsSourceTest.java [hunk 1]
```diff
@@ -9,24 +9,35 @@
...
-import java.time.Instant;
-import java.time.ZoneOffset;
-import org.junit.jupiter.api.AfterEach;
-import org.junit.jupiter.api.BeforeEach;
-import org.junit.jupiter.api.Disabled;
-import org.junit.jupiter.api.Test;
+import net.morimekta.testing.concurrent.FakeClock;
+import org.assertj.core.api.SoftAssertions;
+import org.jetbrains.annotations.NotNull;
+import org.junit.jupiter.api.*;
```
Comment (norrs) [inline java/src/test/java/com/wgtwo/auth/ClientCredentialsSourceTest.java]: I might have to fix my code profile, I guess we don't like wildcard imports?

File: java/src/main/java/com/wgtwo/auth/ClientCredentialSource.java [hunk 2]
```diff
@@ -37,6 +39,6 @@ public BearerTokenCallCredentials callCredentials() {
     private boolean isExpired(Token token) {
         Instant now = clock.instant();
         Instant expiry = token.getExpiry();
-        return now.minusSeconds(120).isAfter(expiry);
+        return now.plus(REFRESH_TOKEN_PERIOD).isAfter(expiry);
```
Comment (norrs) [inline java/src/main/java/com/wgtwo/auth/ClientCredentialSource.java:42]: Bugfix

File: java/src/test/java/com/wgtwo/auth/ClientCredentialsSourceTest.java (hunk 1 above)
Comment (david-wg2) [reply java/src/test/java/com/wgtwo/auth/ClientCredentialsSourceTest.java]: Correct!

File: java/src/test/java/com/wgtwo/auth/ClientCredentialsSourceTest.java [hunk 3]
```diff
@@ -52,22 +63,123 @@ public void shouldCacheUsingClientCredentialsSource() {
...
+        // Act: verify token is recreated within refresh interval or later
+        clock.tick(validPeriods);
+        assertThat(clientCredentialSource.fetchToken().getAccessToken()).isEqualTo(MY_ACCESS_TOKEN_2);
+    }
+
+    static Stream<Arguments> providerOfWithinRefreshIntervalOrLater() {
+        return Stream.of(
+                Arguments.of(named("Within refresh period", WITHIN_REFRESH_PERIOD)),
+                Arguments.of(named("Exactly at refresh period", REFRESH_PERIOD)),
+                Arguments.of(named("'within' life time of a access token but after expiry", Duration.ofMinutes(42))),
```
Comment (norrs) [inline java/src/test/java/com/wgtwo/auth/ClientCredentialsSourceTest.java]:
```suggestion
                Arguments.of(named("'within' life time of an access token but after expiry", Duration.ofMinutes(42))),
```

File: java/pom.xml [hunk 4]
```diff
@@ -22,6 +22,7 @@
         <kotlin.version>1.7.10</kotlin.version>
         <wgtwo.version>1.8.2</wgtwo.version>
         <maven-release-plugin.version>3.0.0-M1</maven-release-plugin.version>
+        <morimekta_utils_testing.version>5.3.0</morimekta_utils_testing.version>
```
Comment (gi-wg2) [inline java/pom.xml]: Use dashes?

File: java/src/main/java/com/wgtwo/auth/ClientCredentialSource.java [hunk 5]
```diff
@@ -12,6 +13,7 @@
  * Will replace the cached token if expired or will expire within two minutes.
  */
 public class ClientCredentialSource {
+    private final static Duration REFRESH_TOKEN_PERIOD = Duration.ofMinutes(2);
```
Comment (gi-wg2) [inline java/src/main/java/com/wgtwo/auth/ClientCredentialSource.java:16]: Nice to have a constant. I am not quite sure about the name, but not sure I can come up with a better one right now...

File: java/src/main/java/com/wgtwo/auth/ClientCredentialSource.java (hunk 2 above)
Comment (gi-wg2) [reply java/src/main/java/com/wgtwo/auth/ClientCredentialSource.java:42]: Nice 🎉

**Review by gi-wg2 — APPROVED** [review]
Nice - Just fix the wildcard, and we'll 🚢

File: java/src/main/java/com/wgtwo/auth/ClientCredentialSource.java (hunk 5 above)
Comment (norrs) [reply java/src/main/java/com/wgtwo/auth/ClientCredentialSource.java:16]: This was actually my second name of the constant. First it was known as jitterSomething, but this is not really about jitter.

File: .gitignore [hunk 6]
```diff
@@ -27,4 +27,3 @@
 hs_err_pid*
 
 target/
-
```
Comment (norrs) [inline .gitignore:30]: pre-commit/ktlint.

**Pull request merged**